### PR TITLE
[NNCFGraph] Migrate from nx.DiGraph to nx.MultiDiGraph

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

* NNCFGraph + InsertionPointGraph + PatternMatchingGraph  are inhereted from `nx.MultiDiGraph` instead of `nx.DiGraph`
* SDPA Ignored pattern is updated with split operation as a branching metatype for OpenVINO/ONNX backends
* Hardware fusing pattern `conv -> arithmetic` is updated to have 1 and 2 degrees edges like on this picture:
image

* ROPE patern is updated to have several parallel input edges to the concat node
image

### Reason for changes

* To make it possible to represent and compress models with Multi edges in NNCF
* To quantize YOLO26 SDPA block properly in OpenVINO/ONNX backends
* To quantize models like a = conv(x); return a*a correctly (not quantizing the x*x part)
* To apply the ROPE pattern correctly, as the original pattern has parallel edges

### Related tickets

174691
179960

### Tests
* MultiDiGraph NNCFGraph is tested here https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-1f470020da6ca32a3f1881ec05b5af9ac828bfd8d892ae1306c860553b960d5bR113-R175
* InsertionPointGraph and PatternMatchingGraph (+SDPA updated ignored pattern +conv+arithmetic HW fused pattern) are tested here https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-7cac922c198eebec6543d9da25eea4fb03aa6e352c4416cf157a41d70010d5feR522-R611
* ONNX MultiDiGraph building test: https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-577b44dd17279c195cfd652ef121c720c5a4278dbcd65cd407932cb7e9d2adccR148-R167
* OpenVINO MultiDiGraph bulding and quantization test: https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-b7a697bf639687bc47899b5166a4e42636c00162cd146739a931382fa1df8a34R1383-R1403
* Torch MultiDIGraph bulding and quantization test: https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-09664d18a9fe689dca4a78998cc8e702526d81b5685c1164b56ce2dcb74b139aR67-R78
* TorchFX MultiDiGraph building test https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-741156c097f1cd42544f5ba4a7dc31024a454b2f78581ebcd2e65f11377fa1a7R80
(Could not quantize yolo26 SDPA subgraph correctly due to the issue https://github.com/openvinotoolkit/nncf/issues/3109)
* Test ROPE pattern with 2 parallel edges: https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-32bc7464d48f05ead87262b533f2e0ce9ea2db101e9e8ae59cb64c65d6aba351R448-R450
* ONNX Synthetic parallel edges model added to check parallel edges nncf graph building: https://github.com/openvinotoolkit/nncf/pull/3843/changes#diff-9aec6e5f328ddb6b440894cb95dcc3b774d34bb67547e9aad6b065bd678c585fR248-R292

### Jobs
Test WC: https://github.com/openvinotoolkit/nncf/actions/runs/21724379639 - Green
Test examples: https://github.com/openvinotoolkit/nncf/actions/runs/21724373555 - Green
PTQ: NNCF/job/manual/job/post_training_quantization/804/ - Green

### CODE REVIEW

File: tests/onnx/quantization/test_transformer_models_graph.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 from tests.onnx.quantization.common import mock_collect_statistics
 from tests.onnx.weightless_model import load_model_topology_with_zeros_weights
 
-TEST_DATA = [ModelToTest("bertsquad-12"), ModelToTest("gpt2-10")]
+TEST_DATA = [ModelToTest("gpt2-10")]
```
Comment (daniil-lyakhov) [inline tests/onnx/quantization/test_transformer_models_graph.py]: Bert squad model present in the repository had 2 edges with the same in_node/out_node and out_port_id/in_port_id, I decided to remove it as consider such ONNX model to be abmiguous and broken

File: src/nncf/common/insertion_point_graph.py [hunk 2]
```diff
@@ -47,13 +47,11 @@ def __str__(self) -> str:
         return self.target_node_name
 
 
-class InsertionPointGraph(nx.DiGraph):  # type: ignore
+class InsertionPointGraph(nx.MultiDiGraph):  # type: ignore
```
Comment (daniil-lyakhov) [inline src/nncf/common/insertion_point_graph.py:50]: Check LSTM sequence workarond is not needed https://github.com/openvinotoolkit/nncf/pull/2802

File: src/nncf/common/graph/graph.py [hunk 3]
```diff
@@ -630,13 +616,10 @@ def get_graph_for_structure_analysis(self, extended: bool = False) -> nx.DiGraph
 
             out_graph.add_node(node_name, **attrs_node)
 
-        for u, v in self._nx_graph.edges:
-            edge = self._nx_graph.edges[u, v]
+        for u, v, k in self._nx_graph.edges:
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply src/nncf/common/graph/graph.py:617]: Is it a valid comment?”

[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply src/nncf/common/graph/graph.py:617]: This is an invalid comment. There is a difference betwen the property `edges` and the function `edges` in the netrokx: the property always returns the tupple of 3 for the MultiDigraph, so this change won't fix anything.
The property call: https://github.com/networkx/networkx/blob/main/networkx/classes/multidigraph.py#L679
 OutMultiEdgeView iter: https://github.com/networkx/networkx/blob/main/networkx/classes/reportviews.py#L1444-L1448

File: src/nncf/common/graph/patterns/patterns.py [hunk 4]
```diff
@@ -191,6 +191,18 @@ def add_pattern_alternative(self, other: "GraphPattern") -> None:
         """
         self._unite_with_copy_of_graph(other.graph)
 
+    def join_patterns_parallel(self, other: "GraphPattern", degree: int) -> None:
```
Comment (AlexanderDokuchaev) [inline src/nncf/common/graph/patterns/patterns.py:194]: Please add tests of new method
And test in test_graph_matching.py to matching graphs with parallel edges

File: src/nncf/common/graph/graph.py [hunk 5]
```diff
@@ -655,15 +642,15 @@ def _get_graph_for_visualization(self) -> nx.DiGraph:
...
+        out_graph = nx.MultiDiGraph()
         for node in self.get_all_nodes():
             attrs_node = {}
             attrs_node["label"] = f"{node.node_id} {node.node_name}"
             node_key = self.get_node_key_by_id(node.node_id)
             out_graph.add_node(node_key, **attrs_node)
 
-        for u, v in self._nx_graph.edges:
-            edge = self._nx_graph.edges[u, v]
+        for u, v, k in self._nx_graph.edges:
```
Comment (AlexanderDokuchaev) [inline src/nncf/common/graph/graph.py:652]: It's can be used as 
```
        for u, v, k, data in self._nx_graph.edges(keys=True, data=True):
```

File: src/nncf/torch/quantization/ignored_patterns.py [hunk 6]
```diff
@@ -246,7 +247,8 @@ def create_rope() -> GraphPattern:
     sin_node = pattern.add_node(**{GraphPattern.LABEL_ATTR: "SIN", GraphPattern.METATYPE_ATTR: om.PTSinMetatype})
 
     pattern.add_edge(matmul_node, transpose_node)
-    pattern.add_edge(transpose_node, concat_node)
+    for _ in range(cat_degree):
```
Comment (AlexanderDokuchaev) [inline src/nncf/torch/quantization/ignored_patterns.py]: As we are discused rope pattern is broken,
Is it really need to change it now? 

And mannually create two edge is shorter and easy to read
```
pattern.add_edge(transpose_node, concat_node)
pattern.add_edge(transpose_node, concat_node)
```
Than 

```
    cat_degree = 2
    ...
    for _ in range(cat_degree):
       pattern.add_edge(transpose_node, concat_node)

File: tests/common/graph/test_nncf_graph.py [hunk 7]
```diff
@@ -107,3 +108,68 @@ def _get_default_nncf_graph_edge(from_node, to_node, input_port_id, output_port_
...
     assert ordinary_edge == output_edges[-1]
+
+
+def test_raise_error_for_duplicated_edge():
+    nncf_graph = NNCFGraph()
+    nodes = []
+    for node in "abc":
+        nodes.append(nncf_graph.add_nncf_node(node, f"type_{node}", f"metatype_{node}"))
+
+    nncf_graph.add_edge_between_nncf_nodes(
```
Comment (AlexanderDokuchaev) [inline tests/common/graph/test_nncf_graph.py]: Looks like edge 0-0 is not used in this test

File: tests/cross_fw/test_templates/helpers.py [hunk 8]
```diff
@@ -494,8 +494,9 @@ def forward(self, x):
 class RoPEModel(nn.Module):
     INPUT_SIZE = [1, 10]
 
-    def __init__(self):
+    def __init__(self, degree: int):
```
Comment (AlexanderDokuchaev) [inline tests/cross_fw/test_templates/helpers.py]: As we was discuesed you should ouse real implementation of rope module
WIth degree=1 it is not valide

File: tests/cross_fw/test_templates/helpers.py [hunk 9]
```diff
@@ -504,7 +505,26 @@ def forward(self, x):
...
         x = torch.matmul(reshape, x)
         x = torch.transpose(x, 2, 1)
-        x = torch.cat([x], dim=2)
+        x = torch.cat([x] * self._degree, dim=2)
+        x1 = x.sin()
+        x2 = x.cos()
+        return x1, x2
+
+
+class RoPEWCModel(nn.Module):
```
Comment (AlexanderDokuchaev) [inline tests/cross_fw/test_templates/helpers.py]: Same it's not valide RoPE module
Please fix it another PR

File: tools/other/render_dot_to_svg.py [hunk 10]
```diff
@@ -28,7 +28,7 @@ def main(argv):
     parser.add_argument("-i", "--input_file", help="Input .dot file", required=True)
     args = parser.parse_args(args=argv)
 
-    graph = nx.DiGraph(read_dot_graph(args.input_file))
+    graph = nx.MultiDiGraph(read_dot_graph(args.input_file))
```
Comment (AlexanderDokuchaev) [inline tools/other/render_dot_to_svg.py:31]: Did you check script from `tools/other`?

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: src/nncf/common/graph/graph.py (hunk 5 above)
Comment (daniil-lyakhov) [reply src/nncf/common/graph/graph.py:652]: Please provide an example when my approach won't work

File: src/nncf/common/graph/patterns/patterns.py (hunk 4 above)
Comment (daniil-lyakhov) [reply src/nncf/common/graph/patterns/patterns.py:194]: Done

File: src/nncf/torch/quantization/ignored_patterns.py (hunk 6 above)
Comment (daniil-lyakhov) [reply src/nncf/torch/quantization/ignored_patterns.py]: I kept them for ONNX/Torch/TorchFX the same but skipped the test, please tell me if you want it in an outher way

File: tests/common/graph/test_nncf_graph.py (hunk 7 above)
Comment (daniil-lyakhov) [reply tests/common/graph/test_nncf_graph.py]: Removed

File: tests/cross_fw/test_templates/helpers.py (hunk 8 above)
Comment (daniil-lyakhov) [reply tests/cross_fw/test_templates/helpers.py]: Removed

File: tests/cross_fw/test_templates/helpers.py (hunk 9 above)
Comment (daniil-lyakhov) [reply tests/cross_fw/test_templates/helpers.py]: Removed

File: tools/other/render_dot_to_svg.py (hunk 10 above)
Comment (daniil-lyakhov) [reply tools/other/render_dot_to_svg.py:31]: Yes, found a bug, fixed and tested manualy, thanks
